### ISSUE TITLE
S3 Artifact multi-namespace security vulnerability

### ISSUE THREAD

**Issue by sander-su** [argoproj/argo#1187]
**Is this a BUG REPORT or FEATURE REQUEST?**:
Both

**What happened**:
Context: 
We want to run workflows in multiple namespaces. 
We want to use s3 for storing artifacts.

Storing and retrieving artifacts is handled by the argoexec sidecar pod. 
For s3 this pod _directly_ connects to s3. 
To connect to s3 the sidecar needs s3 credentials. These credentials need to be present in each namespace that we want to run workflows. 
Having these credentials in each namespace opens up the s3 bucket to all namespaces. Having artifacts from one namespace available to other namespaces. 
Proper namespace isolation cannot be done in this way.

**What you expected to happen**:
Storing and retrieving artifacts is proxied by the controller which handles the connection to s3. The controller first checks if the workflow is authorized to obtain the artifact.
This way direct access to the storage location is shielded and the artifact storage can be used for all namespaces.

Running a controller in each namespace or having a s3 bucket per namespace (currently not possible) are not feasible scaling solutions. 

**How to reproduce it (as minimally and precisely as possible)**:
Run workflows in seperate namespaces

**Anything else we need to know?**:

**Environment**:
- Argo version: 2.2.1
- Kubernetes version : 1.12.4

**Comment by jessesuen** [comment]
In general, I agree that the current implementation of the "default" artifact location in the context of multiple namespaces has some manageability issues. Namely, propagating the secret to all namespaces which needs it, is cumbersome.

However, I consider the current implementation secure -- in that only namespaces which contain the secret credentials, can access the bucket. That said, I agree that there is some management headache of propagating a shared secret to multiple namespace, which I feel needs to be addressed in some shape or form.

> The controller first checks if the workflow is authorized to obtain the artifact.

I think you kind of glossed over this without providing some of the important details, in that what is the mechanism for authorization that would replace the current one?

The current mechanism is that the secret to be created in the namespace which needs to access the bucket. If the secret is not created, the pod will not have access and the workflow will fail. 

I also see some major drawbacks to a proxy approach.

1. By having a centralized proxy where the controller (or some other service) is the one pushing/retrieving artifacts, this service now needs to be reliable and highly available. It is also a potential for a bottleneck and could never scale as well as direct access to AWS S3.

2. By not having a secret per namespace, you lose granular control over what namespaces can or cannot do. For example, you may want one namespace to write to the bucket, and the rest of the namespaces to only read. If this is centralized, then expressing this becomes an entirely different problem.

3. Extending on point 2, namespace granularity may not be sufficient. For example, one use case is to run two workflows with different service accounts in the same namespace. The first workflow can access the bucket, but the second workflow cannot. The current mechanism is able to achieve this by running the pods with different service accounts, with one service account with access to the secret.

In summary, while I think the current approach could use improvement, it should still be considered secure with benefit of being more stable and scalable than an in-cluster proxy based approach.

**Comment by sander-su** [comment]
> However, I consider the current implementation secure -- in that only namespaces which contain the secret credentials, can access the bucket. That said, I agree that there is some management headache of propagating a shared secret to multiple namespace, which I feel needs to be addressed in some shape or form.
The problem here is that it is only possible to configure 1 s3 bucket in the controller. Having a bucket per namespace solve the issue of 

> 
> > The controller first checks if the workflow is authorized to obtain the artifact.
> 
> I think you kind of glossed over this without providing some of the important details, in that what is the mechanism for authorization that would replace the current one?
The controller could check the service account token. This brings it back to the k8s ns isolation. Isolation would than be based on ns or even service account name.
> I also see some major drawbacks to a proxy approach.
> 
>     1. By having a centralized proxy where the controller (or some other service) is the one pushing/retrieving artifacts, this service now needs to be reliable and highly available. It is also a potential for a bottleneck and could never scale as well as direct access to AWS S3.
True, but does it outweigh a simple authorization design. This way a system admin / operator could prevent users from directly accessing the s3 bucket. Furthermore, it would be a stateless proxy which scales excelent.
> 
>     2. By not having a secret per namespace, you lose granular control over what namespaces can or cannot do. For example, you may want one namespace to write to the bucket, and the rest of the namespaces to only read. If this is centralized, then expressing this becomes an entirely different problem.
> 
The controller could ensure no cross namespace reads of subfolders in the bucket. When using service account tokens the service account name could also be the boundry.

>     3. Extending on point 2, namespace granularity may not be sufficient. For example, one use case is to run two workflows with different service accounts in the same namespace. The first workflow can access the bucket, but the second workflow cannot. The current mechanism is able to achieve this by running the pods with different service accounts, with one service account with access to the secret.
Kubernetes isolation is based on namespaces.  When someone requires more isolation this becomes an issue because it is not possible to create an rbac role to specify which service account a user may use in a pod within a namespace. Namespace is the boundery. If one requires more one should use additional namespaces. 
Furthermore, if service account tokens are used for auth to the controller this is also resolved.
 
> 
> In summary, while I think the current approach could use improvement, it should still be considered secure with benefit of being more stable and scalable than an in-cluster proxy based approach.

Slightly disagree, see above.

**Comment by dcherman** [comment]
Just ran across this issue since I'm considering how to secure argo in my own cluster.

As an alternative, couldn't you modify the `keyFormat` in the argo cm to include the workflow namespace as the first part of the path:

https://github.com/argoproj/argo/blob/9ab1bc88f58c551208ce5e76eea0c6fb83359710/docs/workflow-controller-configmap.yaml#L62

`keyFormat: {{workflow.namespace}}/{{workflow.name}}/{{pod.name}}`

Then for each namespace that you could provision a user (or get credentials in another way like IRSA/kiam) that has privs only to the path that corresponds to your namespace.

This wouldn't work for minio since I don't see a way to restrict access to paths and I'm not familiar enough with GCS to determine if it's feasible there, but it should work for S3 directly at least.

**Comment by willzhang** [comment]
i don't want create a new secret, minio have already auto create one , i want use secret in minio namespace, is it possible?
```
root@ubuntu:~# kubectl -n minio get secrets minio -o yaml
apiVersion: v1
data:
  rootPassword: bWluaW8xMjM=
  rootUser: YWRtaW4=
kind: Secret
metadata:
  annotations:
    meta.helm.sh/release-name: minio
    meta.helm.sh/release-namespace: minio
  creationTimestamp: "2023-08-23T08:05:56Z"
  labels:
    app: minio
    app.kubernetes.io/managed-by: Helm
    chart: minio-5.0.13
    heritage: Helm
    release: minio
  name: minio
  namespace: minio
  resourceVersion: "7979291"
  uid: 52146e18-d5ef-4a3e-bc6f-f60b7ee22b97
type: Opaque
```
